geps/gep-2648/index.md

```diff
@@ -0,0 +1,415 @@
+one.
+
+The label is `gateway.networking.k8s.io/policy: direct`.
+
+This solution is intended to allow both users and tooling to identify which CRDs
+in the cluster should be treated as Policy objects, and so is intended to help
+with discoverability generally. It will also be used by the forthcoming `kubectl`
+plugin.
+
+### Conditions
+
+Implementations using Policy objects MUST include a `spec` and `status` stanza,
+and the `status` stanza MUST contain a `conditions` stanza, using the standard
+Condition format.
+
+Policy authors should consider namespacing the `conditions` stanza with a
+`controllerName`, as in Route status, if more than one implementation will be
+reconciling the Policy type.
+
+#### On `Policy` objects
+
+Each Direct Attached Policy MUST populate the `Accepted` condition and reasons
+as defined in the Go spec, a snapshot of which is included below. The canonical
+representation is in the actual Go files. (At the time of writing, this is in
+`apis/v1alpha2/policy_types.go`)
+
+```go
+// PolicyConditionType is a type of condition for a policy.
+type PolicyConditionType string
+
+// PolicyConditionReason is a reason for a policy condition.
+type PolicyConditionReason string
+
+const (
+  // PolicyConditionAccepted indicates whether the policy has been accepted or rejected
+  // by a targeted resource, and why.
+  //
+  // Possible reasons for this condition to be True are:
+  //
+  // * "Accepted"
+  //
+  // Possible reasons for this condition to be False are:
+  //
+  // * "Conflicted"
+  // * "Invalid"
+  // * "TargetNotFound"
+  //
+  PolicyConditionAccepted PolicyConditionType = "Accepted"
+
+  // PolicyReasonAccepted is used with the "Accepted" condition when the policy has been
+  // accepted by the targeted resource.
+  PolicyReasonAccepted PolicyConditionReason = "Accepted"
+
+  // PolicyReasonConflicted is used with the "Accepted" condition when the policy has not
+  // been accepted by a targeted resource because there is another policy that targets the same
+  // resource and a merge is not possible.
+  PolicyReasonConflicted PolicyConditionReason = "Conflicted"
+
+  // PolicyReasonInvalid is used with the "Accepted" condition when the policy is syntactically
+  // or semantically invalid.
+  PolicyReasonInvalid PolicyConditionReason = "Invalid"
+
+  // PolicyReasonTargetNotFound is used with the "Accepted" condition when the policy is attached to
+  // an invalid target resource
+  PolicyReasonTargetNotFound PolicyConditionReason = "TargetNotFound"
+)
+```
+
+#### On targeted resources
+
+Implementations that use Direct Policy objects SHOULD put a Condition into
+`status.Conditions` of any objects affected by a Direct Policy, if that field
+is present.
+
+If they do, that Condition MUST have a `type` ending in `PolicyAffected` (like
+`gateway.networking.k8s.io/PolicyAffected`),
+and have the optional `observedGeneration` field kept up to date when the `spec`
+of the Policy-attached object changes.
```

Comment:
How does this work for implementations like Kuadrant that are implementing policies alongside of the primary implementations of these APIs? I think the intent here is that in the case of a Gateway for example, only the controller implementing that Gateway directly should populate this status?

Reply:
Kuadrant uses `kuadrant.io/DNSPolicyAffected` or `kuadrant.io/TLSPolicyAffected` (depending on the kind of policy), to set the condition on the Gateway object. It also keeps the `observedGeneration` field up to date.

To avoid arm wrestling against the primary gateway controller:
1. it's always additive, relying on [`meta.SetStatusCondition`](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#SetStatusCondition) and [`meta.RemoveStatusCondition`](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#RemoveStatusCondition) - i.e. it tries not to touch on the other conditions;
2. in case of race condition detected and therefore the API server failing to update due to the state of the object changed, it requeues to try again later clean.

Reply:
Thanks @guicassolato, that's exactly what I intended with this, to not block multiple controllers from being able to say "hey, this object is affected by some policy, you should look into that".

Reply:
I added some text above to make this clearer.